mlir/test/Dialect/Linalg/drop-unit-extent-dims.mlir

```diff
@@ -1087,3 +1087,47 @@ func.func @drop_known_unit_constant_low_high(%arg0: tensor<1x383x128xf32>) -> te
 //       CHECK:   } : tensor<383x128xf32> to tensor<384x128xf32>
 //       CHECK:   tensor.expand_shape %[[PADDED]]
 //  CHECK-SAME:     {{\[}}[0, 1], [2]] output_shape [1, 384, 128] : tensor<384x128xf32> into tensor<1x384x128xf32>
+
+
+// -----
+
+// CHECK: #[[$MAP0:.+]] = affine_map<()[s0, s1] -> (s0 * s1)>
+// CHECK: #[[$MAP1:.+]] = affine_map<(d0) -> (0, d0)>
+// CHECK: #[[$MAP2:.+]] = affine_map<(d0) -> ()>
+
+// CHECK-LABEL: func @drop_unit_dim_corresponding_to_dynamic_dim
+// CHECK-SAME:                    %[[ARG0:.*]]: tensor<1x?x?x1xf32>,
+// CHECK-SAME:                    %[[ARG1:.*]]: index) -> tensor<?x1x61x1xf32> {
+// CHECK:           %[[VAL_0:.*]] = arith.constant 0 : index
+// CHECK:           %[[VAL_1:.*]] = arith.constant 1 : index
+// CHECK:           %[[VAL_2:.*]] = arith.constant dense<1.000000e+00> : tensor<f32>
+// CHECK:           %[[VAL_3:.*]] = tensor.collapse_shape %[[ARG0]] {{\[\[}}0, 1], [2, 3]] : tensor<1x?x?x1xf32> into tensor<?x?xf32>
+// CHECK:           %[[VAL_4:.*]] = tensor.empty(%[[ARG1]]) : tensor<?x61xf32>
+// CHECK:           %[[VAL_5:.*]] = affine.apply #[[$MAP0]](){{\[}}%[[ARG1]], %[[VAL_1]]]
+// CHECK:           %[[VAL_6:.*]] = tensor.empty(%[[VAL_5]]) : tensor<?x61xf32>
+// CHECK:           %[[VAL_7:.*]] = linalg.generic {indexing_maps = [#[[$MAP1]], #[[$MAP2]], #[[$MAP1]], #[[$MAP1]]], iterator_types = ["parallel"]} ins(%[[VAL_3]], %[[VAL_2]], %[[VAL_4]] : tensor<?x?xf32>, tensor<f32>, tensor<?x61xf32>) outs(%[[VAL_6]] : tensor<?x61xf32>) {
```

Comment:
To me, this should not happen. The input op had explicitly linked the dimension of one of the operand to the dimension of another. This pattern is dropping that link. The real issue is that the dimensions are inconsistent.

Reply:
Thanks for the review. If I understand correctly you are suggesting that the verifier should error for the `linalg.generic` op since `d0` is inferred to be different for different operands? 

The op is created by the following flow (I am just providing small snippets with relevant ops for brevity):

```
%inserted_slice_37 = tensor.insert_slice %expanded_18 into %15[%c0_25, %c0_28, %c0_30, %c0_31] [1, 1, 61, 1] [1, 1, 1, 1] : tensor<1x1x61x1xf32> into tensor<?x?x?x?xf32>
%18 = linalg.conv_2d_nhwc_hwcf ins(%inserted_slice_37, %cst_2 : tensor<?x?x?x?xf32>, tensor<1x1x1x1xf32>) outs(%17 : tensor<?x1x61x1xf32>) -> tensor<?x1x61x1xf32>
    %reduced = linalg.reduce ins(%cst_7 : tensor<3xi32>) outs(%cst_6 : tensor<i32>) dimensions = [0] 
      (%in: i32, %init: i32) {
        %21 = arith.muli %in, %init : i32
        linalg.yield %21 : i32
      }
```

converts to following after `--canonicalize`:

```
%inserted_slice_13 = tensor.insert_slice %expanded_10 into %7[0, 0, 0, 0] [1, 1, 61, 1] [1, 1, 1, 1] : tensor<1x1x61x1xf32> into tensor<1x?x?x1xf32>
%10 = linalg.conv_2d_nhwc_hwcf ins(%inserted_slice_13, %cst_2 : tensor<1x?x?x1xf32>, tensor<1x1x1x1xf32>) outs(%9 : tensor<?x1x61x1xf32>) -> tensor<?x1x61x1xf32>
  %reduced = linalg.reduce ins(%cst_6 : tensor<3xi32>) outs(%cst_5 : tensor<i32>) dimensions = [0] 
    (%in: i32, %init: i32) {
      %13 = arith.muli %in, %init : i32
      linalg.yield %13 : i32
    }
```

which converts after `--linalg-generalize-named-ops` to : 

```
%8 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2, d3, d4, d5, d6) -> (d0, d1 + d4, d2 + d5, d6)>, affine_map<(d0, d1, d2, d3, d4, d5, d6) -> (d4, d5, d6, d3)>, affine_map<(d0, d1, d2, d3, d4, d5, d6) -> (d0, d1, d2, d3)>], iterator_types = ["parallel", "parallel", "parallel", "parallel", "reduction", "reduction", "reduction"]} ins(%inserted_slice_13, %cst_2 : tensor<1x?x?x1xf32>, tensor<1x1x1x1xf32>) outs(%7 : tensor<?x1x61x1xf32>) {
  ^bb0(%in: f32, %in_16: f32, %out: f32):
    %12 = arith.mulf %in, %in_16 : f32
    %13 = arith.addf %out, %12 : f32
    linalg.yield %13 : f32
  } -> tensor<?x1x61x1xf32>
```

producing the `linalg.genericOp` mentioned in the repro. So, I think in addition to the verifier the canonicalizer should be enhanced to also update the type of the `outs` operand as it is updating the `ins` operand.

Reply:
Looking at the `InferStaticShapeOfOperands` ran during `LinalgOps::canonicalization` the `outs` operand for `linalg.conv_2d_nhwc_hwcf` (and consequently it's `generic` implementation) doesn't get updated because of the check that all maps of `GenericOp` needs to be `projectedPermutation` map. This fails for `#map = affine_map<(d0, d1, d2, d3, d4, d5, d6) -> (d0, d1 + d4, d2 + d5, d6)>` as the second result of the map is not `AffineDimExpr` or `AffineConstExpr` (https://github.com/llvm/llvm-project/blob/d03a1a6e5838c7c2c0836d71507dfdf7840ade49/mlir/lib/IR/AffineMap.cpp#L601). Looking at the rest of the code for updating shapes, I don't follow why this check is present. Do you have any thoughts on why the check is necessary?

Reply:
Skimming through it, I think you are right. Its probably too conservative. Removing that check and might make the static dimensions consistent.

Reply:
New changes 
1. revert the original change in `DropUnitDims`. 
2. removes the `isProjectedPermutation` check in `InferStaticShapeOfOperands`
3. Adds a verifier for `GenericOp` to validate that dimensions are consistent across inputs and outputs operands.

I think the `commonOpVerifier` can be added to any `LinalgOp` but I am not sure what's the best way to do it. Is there a better way than adding the following snippet for all the `LinalgOp`s?

```
LogicalResult <Any>Op::verify() { return commonOpVerifier(*this); }
```

Reply:
Seems like adding the verifier to `verifyStructuredOpInterface` is an option to enable it for all `Linalg` ops. But that caused some failures for existing passes:

`mlir-opt ---pass-pipeline="builtin.module(func.func(tosa-to-linalg-named))"`
```
func.func @matmul_dyn_output(%arg0: tensor<1x1x8xf32>, %arg1: tensor<1x8x1xf32>) -> tensor<?x1x1xf32> {
  %0 = tosa.matmul %arg0, %arg1 : (tensor<1x1x8xf32>, tensor<1x8x1xf32>) -> tensor<?x1x1xf32>
  return %0 : tensor<?x1x1xf32>
}
```

will lead to an error because the `output` dim `0` is dynamic though it should be `1` based on input dims. Running `tosa-infer-shapes` before `tosa-to-linalg-named` solves the problem here but seems like adding the verifier can cause spurious errors like this?

Reply:
Thats unfortunate. I like going down this path. So thanks a lot for doing this. Need to find a way to land this without causing too much friction. I am happy to continue that work.
In the meantime, I will accept your original patch on this. Maybe send that out as a separate PR (and rename this PR?)

Reply:
Thanks, opened https://github.com/llvm/llvm-project/pull/93317 for the original fix.

How are you envisioning fixing the spurious failures? Is it possible to run the `InferStaticShapeOfOperands` as part of building the op?